include/aws/common/array_list.h

```diff
@@ -0,0 +1,126 @@
+    * you want space allocated for. item_size is the size of each element in bytes. Mixing items types is not supported
+    * by this API.
+    */
+    AWS_COMMON_API int aws_array_list_init_dynamic(struct aws_array_list *list,
+        struct aws_allocator *alloc, size_t initial_item_allocation, size_t item_size);
+
+    /**
+    * Initializes an array list with a preallocated array of void *. array_size is the total allocation size of raw_array (not the number of elements),
+    * and item_size is the size in bytes of each element. Mixing items types is not supported
+    * by this API. Note, array_size must be a multiple of item_size or an error will be raised. Once this list is full, new items will be rejected.
+    */
+    AWS_COMMON_API int aws_array_list_init_static(struct aws_array_list *list,
+        void *raw_array, size_t array_size, size_t item_size);
+
+    /**
+    * Deallocates any memory that was allocated for this list, and resets list for reuse or deletion.
+    */
+    AWS_COMMON_API void aws_array_list_clean_up(struct aws_array_list *list);
+
+    /**
+    * Pushes the memory pointed to by val onto the end of internal list
+    */
+    AWS_COMMON_API int aws_array_list_push_back(struct aws_array_list *list, const void *val);
+
+    /**
+    * Returns the element at the front of the list if it exists. If list is empty, AWS_CORE_LIST_ERROR_EMPTY will be raised
+    */
+    AWS_COMMON_API int aws_array_list_front(const struct aws_array_list *list, void *val);
+
+    /**
+    * Deletes the element at the front of the list if it exists. If list is empty, AWS_CORE_LIST_ERROR_EMPTY will be raised.
+    * This call results in shifting all of the elements at the end of the array to the front. Avoid this call unless that is intended
+    * behavior.
+    */
+    AWS_COMMON_API int aws_array_list_pop_front(struct aws_array_list *list);
+
+    /**
+     * Returns the element at the end of the list if it exists. If list is empty, AWS_CORE_LIST_ERROR_EMPTY will be raised.
+     */
+    AWS_COMMON_API int aws_array_list_back(const struct aws_array_list *list, void *val);
+
+    /**
+    * Deletes the element at the end of the list if it exists. If list is empty, AWS_CORE_LIST_ERROR_EMPTY will be raised.
+    */
+    AWS_COMMON_API int aws_array_list_pop_back(struct aws_array_list *list);
+
+    /**
+     * Clears all elements in the array and resets length to zero. Size does not change in this operation.
+     */
+    AWS_COMMON_API void aws_array_list_clear(struct aws_array_list *list);
+
+    /**
+     * If in dynamic mode, shrinks the allocated array size to the minimum amount necessary to store its elements.
+     */
+    AWS_COMMON_API int aws_array_list_shrink_to_fit(struct aws_array_list *list);
+
+    /**
+     * Copies the elements from from to to. If to is in static mode, it must at least be the same length as from.
```

Comment:
We should be clear in documentation (and function naming) as to whether the `to` array is _replaced_ or _appended to_. My instinct is to assume it replaces the `to` array.

Reply:
it should be replacing to, I'll add some docs.